cmd/tool/tool.go

```diff
@@ -24,40 +30,115 @@ func (t CodacyTrivy) Run(tool codacy.Tool, sourceDir string) ([]codacy.Issue, er
 		return []codacy.Issue{}, nil
 	}
 
-	return run(tool.Patterns, tool.Files, sourceDir)
+	config, err := newConfiguration(tool.Patterns, sourceDir)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx := context.Background()
+	return run(ctx, *config)
 }
 
-func run(patterns []codacy.Pattern, files []string, sourceDir string) ([]codacy.Issue, error) {
-	var secretDetectionEnabled = lo.SomeBy(patterns, func(p codacy.Pattern) bool {
-		return p.PatternID == secretRuleID
-	})
-	if !secretDetectionEnabled {
-		return []codacy.Issue{}, nil
+func newConfiguration(patterns []codacy.Pattern, sourceDir string) (*flag.Options, error) {
+	scanners := types.Scanners{}
+	for _, pattern := range patterns {
+		switch pattern.PatternID {
+		case ruleIDSecret:
+			scanners = append(scanners, types.SecretScanner)
+		case ruleIDVulnerability:
+			scanners = append(scanners, types.VulnerabilityScanner)
+		}
+	}
+
+	if len(scanners) == 0 {
+		return nil, ToolError{msg: "Failed to configure Codacy Trivy: no pattern matches existing rules"}
 	}
 
-	scanner := secret.NewScanner(&secret.Config{})
+	// The `quiet` field in global options is not used by the runner.
+	// This is the only way to suppress Trivy logs.
+	log.InitLogger(false, true)
 
-	results := []codacy.Issue{}
+	return &flag.Options{
+		GlobalOptions: flag.GlobalOptions{
+			// CacheDir needs to be explicitly set and match the directory in the Dockerfile.
+			// The cache dir will contain the pre-downloaded vulnerability DBs.
+			CacheDir: cacheDir,
+		},
+		DBOptions: flag.DBOptions{
+			// Do not try to update vulnerability DBs.
+			SkipDBUpdate:     true,
+			SkipJavaDBUpdate: true,
+		},
+		ReportOptions: flag.ReportOptions{
+			// Listing all packages will allow to obtain the line number of a vulnerability.
+			ListAllPkgs: true,
+		},
+		ScanOptions: flag.ScanOptions{
+			// Do not try to connect to the internet to download vulnerability DBs, for example.
+			OfflineScan: true,
+			Scanners:    scanners,
+			// Instead of scanning files individually, scan the whole source directory since it's faster.
+			Target: sourceDir,
+		},
+		VulnerabilityOptions: flag.VulnerabilityOptions{
+			// Only scan libraries not OS packages.
+			VulnType: []types.VulnType{types.VulnTypeLibrary},
+		},
+	}, nil
+}
+
+func run(ctx context.Context, config flag.Options) ([]codacy.Issue, error) {
+	runner, err := artifact.NewRunner(ctx, config)
+	if err != nil {
+		return nil, ToolError{msg: "Failed to initialize Codacy Trivy", w: err}
+	}
+	defer runner.Close(ctx)
+
+	results, err := runner.ScanFilesystem(ctx, config)
```

Comment:
hmm, this means we're going to be supporting the same languages as secret detection
given that for now we are returning "all good!" for some languages when they don't support line number, Codacy is going to cache those files, and ignore them when we add support for it

a workaround would be to return a `FileError` in order to invalidate their cache. the consequence is that Codacy would try to run them at every analysis (https://github.com/codacy/codacy-engine-golang-seed/blob/161d02ec4bcbbd4d9afab9fe7381b9a98bd11387/error.go#L8)